# 4.6.2 Combining & Exporting Data

### This script contains the following points:

#### 01. Importing Libraries
#### 02. Importing Data & Checking Dimensions
#### 03. Combining the Orders_Products_Combined DataFrame with the Cleaned Products Dataset
#### 04. Confirm Results of Merge Using Merge Flag
#### 05. Exporting in Pickle Format

## 01. Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 02. Importing Data & Checking Dimensions

In [2]:
# Turn Path into a String & Assign Python Shortcut
path = '/Users/mgottlieb/Documents/Instacart Basket Analysis'

In [3]:
# Import the orders_products_combined.pkl dataset
df_ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [4]:
#Checking the output of the import
df_ords_prods_combined.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id evaluation_status  \
0             0           0   2539329        1             prior   
1             0           0   2539329        1             prior   
2             0           0   2539329        1             prior   
3             0           0   2539329        1             prior   
4             0           0   2539329        1             prior   

   order_number  orders_day_of_week  order_hour_of_day  \
0             1                   2                  8   
1             1                   2                  8   
2             1                   2                  8   
3             1                   2                  8   
4             1                   2                  8   

   days_since_prior_order  days_since_prior_order_missing  product_id  \
0                     7.0                               1         196   
1                     7.0                               1       14084   
2                     7.0                               1       12427   
3                     7.0                               1       26088   
4                     7.0                               1       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In [5]:
# Droping the two Unnamed columns from the Dataframe (altering the dataframe)
df_ords_prods_combined.drop(columns=['Unnamed: 0.1', 'Unnamed: 0'], inplace=True)

In [6]:
#Checking the output with this modification
df_ords_prods_combined.head()

order_id  user_id evaluation_status  order_number  orders_day_of_week  \
0   2539329        1             prior             1                   2   
1   2539329        1             prior             1                   2   
2   2539329        1             prior             1                   2   
3   2539329        1             prior             1                   2   
4   2539329        1             prior             1                   2   

   order_hour_of_day  days_since_prior_order  days_since_prior_order_missing  \
0                  8                     7.0                               1   
1                  8                     7.0                               1   
2                  8                     7.0                               1   
3                  8                     7.0                               1   
4                  8                     7.0                               1   

   product_id  add_to_cart_order  reordered _merge  
0         196                  1          0   both  
1       14084                  2          0   both  
2       12427                  3          0   both  
3       26088                  4          0   both  
4       26405                  5          0   both

In [7]:
df_ords_prods_combined.shape

(32434489, 12)

In [8]:
# Import dataset products_checked.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

In [9]:
#Checking the output of the import
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [10]:
df_prods.shape

(49672, 6)

In [11]:
# Droping the Unnamed column from the Dataframe (altering the dataframe)
df_prods.drop(columns=['Unnamed: 0'], inplace=True)

In [12]:
#Checking the output with this modification
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [13]:
df_prods.shape

(49672, 5)

## 03. Combining the Orders_Products_Combined DataFrame with the Cleaned Products Dataset

First I need to determine the key (column) that exists in both datasets 'product_id', for this project I am just using the default inner joined.

In [14]:
df_ords_prods_merge = pd.merge(df_ords_prods_combined, df_prods, on='product_id',)

In [15]:
#Checking the output of the merge
df_ords_prods_merge.head()

order_id  user_id evaluation_status  order_number  orders_day_of_week  \
0   2539329        1             prior             1                   2   
1   2398795        1             prior             2                   3   
2    473747        1             prior             3                   3   
3   2254736        1             prior             4                   4   
4    431534        1             prior             5                   4   

   order_hour_of_day  days_since_prior_order  days_since_prior_order_missing  \
0                  8                     7.0                               1   
1                  7                    15.0                               0   
2                 12                    21.0                               0   
3                  7                    29.0                               0   
4                 15                    28.0                               0   

   product_id  add_to_cart_order  reordered _merge product_name  aisle_id  \
0         196                  1          0   both         Soda        77   
1         196                  1          1   both         Soda        77   
2         196                  1          1   both         Soda        77   
3         196                  1          1   both         Soda        77   
4         196                  1          1   both         Soda        77   

   department_id  prices  
0              7     9.0  
1              7     9.0  
2              7     9.0  
3              7     9.0  
4              7     9.0

In [16]:
df_ords_prods_merge.shape

(32404859, 16)

## 04. Confirm Results of Merge Using Merge Flag

In [17]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

The _merge output indicates that all 32,404,859 rows from the merged DataFrame were matched entries present in both original data sets (both), with no unmatched entries from either the left (left_only) or right (right_offset) data sets.

## 05. Exporting in Pickle Format

Since I know these files will be used in Python, I choice Pickle since it can be imported and exported quickly, I know that the reimported dataframe won’t be changed, and the a high compression rate.

In [19]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))